src/main/java/com/scalar/db/storage/cosmos/ResultImpl.java

```diff
@@ -0,0 +1,157 @@
+package com.scalar.db.storage.cosmos;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.MoreObjects;
+import com.scalar.db.api.Result;
+import com.scalar.db.api.Selection;
+import com.scalar.db.exception.storage.InvalidMetadataException;
+import com.scalar.db.exception.storage.UnsupportedTypeException;
+import com.scalar.db.io.BigIntValue;
+import com.scalar.db.io.BlobValue;
+import com.scalar.db.io.BooleanValue;
+import com.scalar.db.io.DoubleValue;
+import com.scalar.db.io.FloatValue;
+import com.scalar.db.io.IntValue;
+import com.scalar.db.io.Key;
+import com.scalar.db.io.TextValue;
+import com.scalar.db.io.Value;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.Immutable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Immutable
+public class ResultImpl implements Result {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ResultImpl.class);
+  private final TableMetadata metadata;
+  private final Map<String, Value> values;
+
+  public ResultImpl(Record record, Selection selection, TableMetadata metadata) {
+    checkNotNull(record);
+    this.metadata = checkNotNull(metadata);
+    values = new HashMap<>();
+    interpret(record, selection, metadata);
+  }
+
+  @Override
+  public Optional<Key> getPartitionKey() {
+    return getKey(metadata.getClusteringKeyNames());
+  }
+
+  @Override
+  public Optional<Key> getClusteringKey() {
+    return getKey(metadata.getClusteringKeyNames());
+  }
+
+  @Override
+  public Optional<Value> getValue(String name) {
+    return Optional.ofNullable(values.get(name));
+  }
+
+  @Override
+  @Nonnull
+  public Map<String, Value> getValues() {
+    return Collections.unmodifiableMap(values);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(values);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (o == this) {
+      return true;
+    }
+    if (!(o instanceof ResultImpl)) {
+      return false;
+    }
+    ResultImpl other = (ResultImpl) o;
+    if (this.values.equals(other.values)) {
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this).add("values", values).toString();
+  }
+
+  private void interpret(Record record, Selection selection, TableMetadata metadata) {
+    Map<String, Object> recordValues = new HashMap<>();
+    recordValues.putAll(record.getPartitionKey());
```

Comment:
I think copying to `recordValues` can be avoided by having a function like this. (the
What do you think ? (Sorry the indent isn't adjusted)

```
record.getPartitionKey().forEach((name, value) -> add(values, name, value));
record.getClusteringKey().forEach((name, value) -> add(values, name, value));
...

selection
          .getProjections()
          .forEach(
              name -> {
                if (onlyValues.containsKey(name)) {
                  add(name, record.getValues().get(name));
                }
              });
    }


private void add(Map<String, Value> values, String name, Object value) {
          if (metadata.getColumns().containsKey(name)) {
            values.put(name, convert(value, name, metadata.getColumns().get(name)));
          } else {
            throw new InvalidMetadataException(
                "metadata doesn't have the specified column: " + name);
          }
        });
}
```

Reply:
Thank you! Modified!